### DISCUSSION TITLE
Comparing ASAP7 and sky130hd with respect to ALU operations

### DISCUSSION THREAD

**Discussion by oharboe** [The-OpenROAD-Project/OpenROAD#3881] [General]
The [mock-alu](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/tree/master/flow/designs/src/mock-alu) allows studying the speed and area of various typical ALU operations.

The mock-alu has two 64 bit inputs and one 64 bit output. The inputs and outputs are registered within the mock-alu. This is to have clear boundary conditions when studying the guts of the mock-alu.

The mock-alu implements a number of operations and variants of these operations to study the area and minimum clock period for these operations.

The operations fall into a couple of categories:

- *add, subtract and compare*: ADD, SUB, SETCC_EQ(equal), SETCC_NE (not equal), SETCC_LT (less than), SETT_LE (less than or equal), SETCC_ULT (unsigned less than), SETCC_ULE(unsigned less than or equal)
- *barrel shifter*: SHR(logical shift right), SRA(arithmetic shift right), SHL(shift left)
- *bitwise logic*: OR(bitwise or), AND(bitwise and), XOR(bitwise xor)
- *multiplication*: 64 bit multiply. There are various algorithms used, default Han Carlson. The implementation is PDK specific and comes from https://github.com/antonblanchard/vlsiffra/
- *multiplexor*: MUX1..8. This is not really an ALU operation. All that is happening here is that bits from the input as selected using a mux and put into the output. This allows studying the performance of the mux that sits before the output of an ALU, which is selecting between the various supported operations.

Next, the mock-alu allows implementing any combination of these operations. This allows implementing an ALU that only supports the shift operations, which can be labelled "SHR,SHL,SRA". This shift operation only mock-alu has a single shared barrelshifter. Similarly, a bitwise logic only mock-alu, can be labelled "OR,AND,XOR".

At this point, we can plot various mock-alu implementations for ASAP7 and sky130hd:

![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/2798822/b9234684-43f4-4a74-98dd-81cf8b74f9ce)

![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/2798822/fefae238-58ed-4bfe-a307-780046acdb83)

Here 8,16,32 and 64 wide ADD operations are plotted:

![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/2798822/f4a837e6-700c-4436-81ac-221bdfafdb57)

Various multiplication algorithms for 64 bit multiplication, 4 pipeline stages:

![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/2798822/0a8d6518-5999-4d18-87fa-38a13857764f)

Plotting Han Carlson multiply algorithm with 8, 16, 32 and 64 bit bit width:

![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/2798822/97fb01d7-e535-4578-9fe9-cdb80a4e945f)

Thoughts?

**Comment by maliberty** [comment]
Its interesting that the various multiply algorithms have the same delay (excluding ripple).

I don't see a MULT for sky130

**Reply by oharboe** [reply]
Indeed. I was also surprised about the minescule difference in area for the algorithms.

I have not set up multiply for sky130hd yet.I was more interested in the simpler ALU operations.

**Reply by oharboe** [reply]
![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/2798822/f2c298f9-fe0f-4a5f-8012-17946ef478cb)

**Reply by oharboe** [reply]
Something is fishy... Why is the minimum clock period very nearly identical?

```
make DESIGN_CONFIG=designs/asap7/mock-alu/config.mk FLOW_VARIANT=MULT_KOGGESTONE PLATFORM=sky130hd gui_route
```

```
>>> report_checks -path_delay max -fields {fanout, cap, slew} -format full_clock
Startpoint: _0351_ (rising edge-triggered flip-flop clocked by clock)
Endpoint: _0255_ (rising edge-triggered flip-flop clocked by clock)
Path Group: clock
Path Type: max

    Cap    Slew   Delay    Time   Description
-----------------------------------------------------------------------
                   0.00    0.00   clock clock (rise edge)
                   0.00    0.00   clock source latency
   0.08    0.00    0.00    0.00 ^ clock (in)
   0.15    0.45    0.41    0.41 ^ clkbuf_0_clock/X (sky130_fd_sc_hd__clkbuf_4)
   0.08    0.25    0.40    0.81 ^ clkbuf_3_0__f_clock/X (sky130_fd_sc_hd__clkbuf_4)
   0.09    0.28    0.38    1.19 ^ clkbuf_leaf_60_clock/X (sky130_fd_sc_hd__clkbuf_4)
           0.28    0.00    1.19 ^ _0351_/CLK (sky130_fd_sc_hd__dfxtp_1)
   0.01    0.05    0.39    1.57 v _0351_/Q (sky130_fd_sc_hd__dfxtp_1)
   0.01    0.10    0.51    2.09 v multResult_mult.dadda_fa_0_61_0/SUM (sky130_fd_sc_hd__fa_1)
   0.02    0.11    0.57    2.65 v multResult_mult.dadda_fa_1_61_8/COUT (sky130_fd_sc_hd__fa_4)
   0.01    0.10    0.52    3.17 v multResult_mult.dadda_fa_2_62_3/SUM (sky130_fd_sc_hd__fa_1)
   0.01    0.07    0.53    3.71 ^ multResult_mult.dadda_fa_3_62_3/SUM (sky130_fd_sc_hd__fa_1)
   0.01    0.09    0.43    4.14 v multResult_mult.dadda_fa_4_62_2/SUM (sky130_fd_sc_hd__fa_1)
   0.01    0.09    0.40    4.54 v multResult_mult.dadda_fa_5_62_1/COUT (sky130_fd_sc_hd__fa_1)
   0.01    0.08    0.54    5.07 ^ multResult_mult.dadda_fa_6_63_0/SUM (sky130_fd_sc_hd__fa_1)
   0.00    0.07    0.40    5.48 v multResult_mult.dadda_fa_7_63_0/SUM (sky130_fd_sc_hd__fa_1)
           0.07    0.00    5.48 v _0255_/D (sky130_fd_sc_hd__dfxtp_1)
                           5.48   data arrival time

                   5.00    5.00   clock clock (rise edge)
                   0.00    5.00   clock source latency
   0.08    0.00    0.00    5.00 ^ clock (in)
   0.15    0.45    0.41    5.41 ^ clkbuf_0_clock/X (sky130_fd_sc_hd__clkbuf_4)
   0.08    0.25    0.40    5.81 ^ clkbuf_3_0__f_clock/X (sky130_fd_sc_hd__clkbuf_4)
   0.07    0.22    0.33    6.14 ^ clkbuf_leaf_61_clock/X (sky130_fd_sc_hd__clkbuf_4)
           0.22    0.00    6.14 ^ _0255_/CLK (sky130_fd_sc_hd__dfxtp_1)
                   0.00    6.14   clock reconvergence pessimism
                  -0.08    6.06   library setup time
                           6.06   data required time
-----------------------------------------------------------------------
                           6.06   data required time
                          -5.48   data arrival time
-----------------------------------------------------------------------
                           0.58   slack (MET)
```


![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/2798822/9c36c146-3f03-4fcd-8831-7a03a704e6d3)

**Reply by maliberty** [reply]
What do you mean by "the time period"?  Identical to what?

**Reply by oharboe** [reply]
The minimum clock period(corrected above) hardly varies across algorithms, except for ripple, which is much slower, as expected.

**Reply by maliberty** [reply]
The same as my comment "Its interesting that the various multiply algorithms have the same delay (excluding ripple)."

**Reply by QuantamHD** [reply]
If you're interested you might want to try a booth multiplier from https://github.com/antonblanchard/vlsiffra

**Reply by oharboe** [reply]
booth isn't mentioned in the docs on that link, nor can I see it in the command line parser.

I missed it when I first looked, but I am going to try out vlsi-adder.

**Reply by QuantamHD** [reply]
> vlsiffra achieves this by using many well established techniques including [Booth encoding](https://en.wikipedia.org/wiki/Booth%27s_multiplication_algorithm), [Dadda reduction](https://en.wikipedia.org/wiki/Dadda_multiplier) and a choice of fast adders like [Kogge-Stone](https://en.wikipedia.org/wiki/Kogge%E2%80%93Stone_adder).

**Reply by oharboe** [reply]
I see. As I understand booth is used within the algorithms that are listed in `vlsi-multiplier` algorithms and I have plotted those results above for these multiplication algorithms.

**Comment by oharboe** [comment]
Here are the `vlsi-adder` algorith, excluding ripple that was ca. 2000+ ps. It would have made the graph harder to read:

![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/2798822/07c8e764-81d1-48fe-858a-39175a1d3c68)

**Reply by oharboe** [reply]
With negative slack and TNS_END_PERCENT=100:

![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/2798822/798253aa-9657-45fe-bcef-838ceaafdb7e)

**Reply by maliberty** [reply]
A bit mixed.  Some are improved and others are just bigger.

**Reply by oharboe** [reply]
Negative slack for multipliers, 64 bit, 4 pipeline stages.

This is a great example of how one should not be fooled by the scales on the axis....

There's essentially no difference between these implementations.

Leaving out MULT_RIPPLE as it is far, far off to the right and the bottom. As expected.


![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/2798822/10ba9862-02dc-46ab-ac2f-25bead027a7b)

**Comment by maliberty** [comment]
Do all your runs end in negative slack at the end of the cts & the complete flow?  Once we reach zero slack we stop optimizing.

**Reply by oharboe** [reply]
Ah, I see. No, they have positive slack.

**Comment by maliberty** [comment]
One thing worth noting is that asap7 doesn't have a full or half adder cells in the library (sky130 does).

**Reply by oharboe** [reply]
Can you explain a bit more what that means.

I saw full adders and half adders here...

https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/blob/7db5d2c1fc7a2d99711ea3af19c9567dd45e371e/flow/platforms/asap7/yoSys/cells_adders_L.v#L28

**Reply by maliberty** [reply]
I take it back - it does seem to have them.  The $fa is the full adder in yosys and this code maps it to the tech dependent cells (if you have a constant on a full adder then you can use a half adder instead).

**Comment by mithro** [comment]
BTW Have you seen Teo's spreadsheet @ https://docs.google.com/spreadsheets/d/1pTGzZo5XYU7iuUryxorfzJwNuE9rM3le5t44wmLohy4/edit#gid=126548956 ?

**Reply by oharboe** [reply]
Fascinating. So much to explore on simple addition :-)

**Reply by mithro** [reply]
FYI, Here is a link dump of the various stuff around adders that I have seen (in case you haven't seen these before);

#### Teo's work

- https://docs.google.com/presentation/d/1A8KDiTH1qyAzy9gKalUJgxLuPrOgqfOg7t7j_XdqeGE/edit#slide=id.p1
- https://www.youtube.com/watch?v=ZvssTXMTRmk
- https://github.com/tdene/synth_opt_adders
- https://github.com/tdene/adder_networks
- [Skywater Adder Performance Spreadsheet](https://docs.google.com/spreadsheets/d/1pTGzZo5XYU7iuUryxorfzJwNuE9rM3le5t44wmLohy4/edit#gid=126548956)
- [SKY130 Maximum clock frequency (performance / speed) for DFF](https://docs.google.com/spreadsheets/d/1BTvnNFhkFqT4uqZs7uIK6D6QAJ4tRYkpU44OZFgme9Q/edit#gid=1572495657)

#### Matt Venn testing of adders

- https://github.com/mattvenn/instrumented_adder
- https://github.com/mattvenn/wrapped_instrumented_adder
- https://www.youtube.com/watch?v=O38i5Y98m44
- https://docs.google.com/presentation/d/12Errw5M7lFYbBfi9yHc9-JNaikUcMz3qiLRNhj5hiUc/edit#slide=id.p
- https://www.youtube.com/watch?v=Gg6mxvhiEUs <-- Matt Venn mentions having gotten silicon back

#### Other stuff

It looks like you have already seen Anton's work @ https://github.com/antonblanchard/vlsiffra

Also checkout Alan Mishchenko's (the author of ABC) recent pre-prints for yet another direction in this space.

I'm sure I'm missing a bunch of stuff too!

**Comment by mithro** [comment]
BTW I would love to get a similar spreadsheet to Teo's for GF180MCU and ASAP7 too. 

Sadly, Teo got distracted by the mathematical theory and then was stolen by NVIDIA before he could get to that.

**Comment by tspyrou** [comment]
@oharboe did you try using set_clock_uncertainty to force the slack to be negative and make the tool work harder.
As @maliberty mentioned optimization will stop once timing passes.

**Reply by oharboe** [reply]
I simply reduced the clock period to have negative slack and set TNS_END_PERCENT to 100. It changed the results a bit for the better, but I didnt see any dramatic change in relationships between algorithms.

**Comment by mithro** [comment]
@oharboe - Any chance you could do a write up of what you discovered?

**Reply by oharboe** [reply]
It is in this post... Any particular questions that come to mind that I could clarify?

**Reply by mithro** [reply]
There is a lot bunch of back and forth and I'm unclear what the final results are (and exactly how you produced them). 

Various questions include;
 * How do I reproduce your results and graphs?
 * What settings did you end up using and why did you end up using those settings? (Particularly around making the tool work harder?)
 * How do the various implementations compare? Do you understand why the compare this way?
 * How do SKY130 and ASAP7 compare in the end here?
 * Do the relative "positions" between the implementations hold across SKY130 and ASAP7?
 * What was the most interesting / unexpected thing you discovered?

Writing it up as a nice coherent blog post would be pretty awesome but totally understand if you do not have the time to do so.

**Comment by oharboe** [comment]
> There is a lot bunch of back and forth and I'm unclear what the final results are (and exactly how you produced them).
> 
> Various questions include;
> 
> * How do I reproduce your results and graphs?

Run this script:  https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/blob/master/flow/designs/src/mock-alu/plot-area-min-clock-period.py

Some tinkering required.

> * What settings did you end up using and why did you end up using those settings? (Particularly around making the tool work harder?)

I didn't study how to make a best possible ALU, I was only interested the relationship between the ALU operations.

> * How do the various implementations compare? Do you understand why the compare this way?

The various implementation of additions and multiplications?

It is a mystery why there is essentially no difference between multiplication implementation clock periods...

> * How do SKY130 and ASAP7 compare in the end here?

The lessons learned on the relative size and speed of ALU operations appear to be much the same with SKY130 and ASAP7. Which is surprising. Ca. 15 years separate them...

> * Do the relative "positions" between the implementations hold across SKY130 and ASAP7?

Pretty much.

> * What was the most interesting / unexpected thing you discovered?

That relative size and speed of simple ALU operations are essentially unchanged across process nodes.

Also, it would appear that if an ALU operation is 200ps on x86 7nm, yielding 5GHz, then one could choose to divide clock period of ASAP7+OpenROAD by 4 for simple ALU operations when one models and decide to take the lessons learned and apply them to architectural exploration. Further choose to treat the speed of ASAP7+OpenROAD as not terribly important in terms of making architctural choices as the choices will be the same if everything is optimized.

By this I mean that to drive your architctural exploration, as a [first order approximation](https://en.wikipedia.org/wiki/Order_of_approximation#First-order), write the RTL in an idiomatic way, run them through ASAP7+OpenROAD and if your design is 4x the desired clock period, your design isn't completely off.

Nobody who have information on commercial tools and PDKs can challenge me here. :-) Not because I'm right, but because PDKs and commercial tools are under strict NDAs... This also explains why there are a lot of unsaid things in this thread...

> Writing it up as a nice coherent blog post would be pretty awesome but totally understand if you do nt have the time to do so.

Agreed. At least I summarize a bit here. I'm happy to hear that there are some that are interested in this.

Perhaps you would like to write a blog-post using the script above?
